src/Raven.Server/ServerWide/ServerStore.cs

```diff
@@ -182,6 +184,21 @@ public ServerStore(RavenConfiguration configuration, RavenServer server)
 
             if (Configuration.Indexing.MaxNumberOfConcurrentlyRunningIndexes != null)
                 ServerWideConcurrentlyRunningIndexesLock = new FifoSemaphore(Configuration.Indexing.MaxNumberOfConcurrentlyRunningIndexes.Value);
+
+            CatastrophicFailureNotification = new CatastrophicFailureNotification((envId, path, exception, stacktrace) =>
+            {
+                try
+                {
+                    if (Logger.IsOperationsEnabled)
+                        Logger.OperationsAsync($"Catastrophic failure in server storage located at '{path}', StackTrace:'{stacktrace}'", exception).Wait(TimeSpan.FromSeconds(1));
+                }
+                catch
+                {
+                    // nothing we can do
+                }
+
+                Environment.Exit(29); // ERROR_WRITE_FAULT
```

Comment:
@gregolsky is this enough for services to restart? There is also `FailFast` method. According to remarks it would be a better 'fit': https://docs.microsoft.com/en-us/dotnet/api/system.environment.failfast?view=net-5.0

Reply:
`FailFast` means that it will skip things like finalizers, no reason to do that, however, in this case, I think.

Reply:
Yes, non-zero process exit code will cause the service to restart - on Linux we have restart on failure statement in the service file: https://github.com/ravendb/ravendb/blob/v5.2/scripts/linux/ravendb.service#L14